### PR TITLE
Removed the DefaultValue attribute from ListBox.ItemHeight property

### LINKED ISSUE #4463: The code generated by the ListBox control is not consistent with Framework
* .NET Core Version: 
3.1+

* Have you experienced this same bug with .NET Framework?: 
No

**Problem description:**

The generated code for ListBox1 have extra code: this.listBox1.ItemHeight = 15;
image

**Expected behavior:**
The generated code for ListBox1 should not have this code: this.listBox1.ItemHeight = 15;
image

**Minimal repro:**
1. Create a Core winforms application.
2. Add a **ListBox** control to **form**.
3. Double click the **Form1.Designer.cs** file in Solution Explorer.
4. Observe the generated code of the **ListBox**.
Result: The generated code for ListBox1 should not have this code: this.listBox1.ItemHeight = 15;

Note from Dustin: ListBox.ItemHeight property. 
dotnet/winforms@4bca5b3/src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs#L596-L616
It has a `[DefaultValue(DefaultItemHeight)]` attribute and DefaultItemHeight is 13.
Note from Tanya: Our default value is not accurate anymore due to the default font change. 13 is the default for Microsoft Sans Serif, 8.25pt, 15 is what comctl32 listbox calculates for Segoe UI, 9pt. To confirm - change the parent control's font.

### PR DESCRIPTION
**Pull request by dkazennov**
Fixes #4463 (partially, see also #8640)

## Proposed changes

- Introduces a new private property `DefaultHeight` with **the intended value** equals Winforms Control.DefaultFont.Height (for example in Windows.Forms it is **Segoe UI, Size 9, Height 16** by default).
- Removes the DefaultValue attribute from the `ListBox.ItemHeight` property;
- Implements `ResetItemHeight` and `ShouldSerializeItemHeight` methods ([see more](https://learn.microsoft.com/en-us/dotnet/desktop/winforms/controls/defining-default-values-with-the-shouldserialize-and-reset-methods?view=netframeworkdesktop-4.8)).
- The ListBox.DefaultItemHeight constant marked as obsolete. The `ListBox.ItemHeight` is now depended on actual `DefaultFont.Height` value.
- The `ListBox.ItemHeight` property still returns the value of the  `_itemHeight` property as usual.
- Changed 9 ListBox unit tests

## Customer Impact

- Fixes the Issue which impacts user experience,  breaks DPI and Font scaling. For the _newly created_ control at 100% scaling, designer should generate the same value as is defined by the Default. There is no extra code.

## Regression? 

- No

## Risk

- Minimal

### Before

image

### After

image

## Test methodology

- Manual (Issue can be tested via substitution of dlls, complete restart of the VS recommended)
- Unit test

 

## Test environment(s) 

- dotnet 8.0.100-alpha.1.22607.6

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8518)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 1]
```diff
@@ -51,6 +51,7 @@ public partial class ListBox : ListControl
         /// <summary>
         ///  The default item height for an owner-draw ListBox. The ListBox's non-ownerdraw
         ///  item height is 13 for the default font on Windows.
+        ///  This constant is obsolete.
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:54]: To properly obsolete a member, you need to use `ObsoleteAttribute`, for example: https://github.com/dotnet/winforms/blob/37e19459f4554e4b69de9cb5f95d623d272c899d/src/System.Windows.Forms/src/System/Windows/Forms/DomainUpDown.DomainItemAccessibleObject.cs#L12-L17
And follow the process described in https://github.com/dotnet/winforms/blob/main/docs/list-of-diagnostics.md

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 2]
```diff
@@ -389,6 +390,11 @@ public bool UseCustomTabOffsets
             }
         }
 
+        private static int DefaultHeight
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:393]: Name `ListBox.DefaultHeight` sounds like it refers to the whole `ListBox`, rather than to the height of a single item. What do you think about `DefaultItemHeightInternal`?

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:393]: `Internal` implies that there is another property for external use. We are actually deprecating that other `const`.  Though I can't come up with a great name... `DefaultListBoxItemHeight`?

**Review by NikitaSemenovAkvelon — APPROVED** [review]
LGTM

**Comment by dkazennov** [comment]
Some builds failed due to issues with integration tests unrelated to this fix (and those test are not shown in the list of tests in the pipeline description).

File: src/Common/src/Obsoletions.cs [hunk 3]
```diff
@@ -21,4 +21,9 @@ internal static class Obsoletions
     internal const string DomainItemAccessibleObjectMessage = $"{nameof(DomainUpDown.DomainItemAccessibleObject)} is no longer used to provide accessible support for {nameof(DomainUpDown)} items.";
 #pragma warning restore WFDEV003 // Type or member is obsolete
     internal const string DomainItemAccessibleObjectDiagnosticId = "WFDEV003";
+
+#pragma warning disable WFDEV004
+    internal const string ListBoxDefaultItemHeightMessage = $"The {nameof(ListBox.DefaultItemHeight)} const is replaced with the ListBox.DefaultListBoxItemHeight property.";
```
Comment (dmitrii-drobotov) [inline src/Common/src/Obsoletions.cs:26]: Referring to a private property may not be helpful for those who will want to replace the const in their code. I suggest mentioning what can be used instead and also saying that default height is now dependent on application's font.

```suggestion
    internal const string ListBoxDefaultItemHeightMessage = $"{nameof(ListBox.DefaultItemHeight)} is no longer used as default item height, which now depends on application's default font. Use {nameof(ListBox.ItemHeight)} instead.";
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 4]
```diff
@@ -389,6 +394,11 @@ public bool UseCustomTabOffsets
             }
         }
 
+        private static int DefaultListBoxItemHeight
+        {
+            get => DefaultFont.Height;
+        }
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:400]: Nit:
```suggestion
        private static int DefaultListBoxItemHeight => DefaultFont.Height;
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs [hunk 5]
```diff
@@ -1382,6 +1382,30 @@ public static IEnumerable<object[]> HorizontalExtent_Set_TestData()
             }
         }
 
+        [WinFormsFact]
+        public void ListBox_Height_IsEqualDefaultHeight_IfDefaultFontIsChanged()
+        {
+            var oldApplicationDefaultFont = Application.DefaultFont;
+            using var font = new Font("Times New Roman", 12);
+            using var listBox = new TestListBox();
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs:1390]:
```suggestion
            using Font font = new("Times New Roman", 12);
            using TestListBox listBox = new();
```

**Review by dmitrii-drobotov — COMMENTED** [review]
Can you also update the table in https://github.com/dotnet/winforms/blob/main/docs/list-of-diagnostics.md in this PR?

File: src/Common/src/Obsoletions.cs (hunk 3 above)
Comment (dkazennov) [reply src/Common/src/Obsoletions.cs:26]: Yes, that's a good point here.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs [hunk 6]
```diff
@@ -1382,6 +1382,30 @@ public static IEnumerable<object[]> HorizontalExtent_Set_TestData()
...
+            using Font font = new Font("Times New Roman", 12);
+            using TestListBox listBox = new TestListBox();
+            var applicationTestAccessor = typeof(Application).TestAccessor().Dynamic;
+
+            try
+            {
+                applicationTestAccessor.s_defaultFont = font;
+                Assert.NotEqual(oldApplicationDefaultFont, Application.DefaultFont);
+
+                var rowDefaultHeight = listBox.GetDefaultListBoxItemHeight();
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs:1398]: Since `GetDefaultListBoxItemHeight` is used only here, perhaps it can be inlined and then TestListBox can be removed?
```suggestion
                int rowDefaultHeight = listBox.TestAccessor().Dynamic.DefaultListBoxItemHeight;
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs [hunk 7]
```diff
@@ -1382,6 +1382,30 @@ public static IEnumerable<object[]> HorizontalExtent_Set_TestData()
             }
         }
 
+        [WinFormsFact]
+        public void ListBox_Height_IsEqualDefaultHeight_IfDefaultFontIsChanged()
+        {
+            var oldApplicationDefaultFont = Application.DefaultFont;
+            using Font font = new Font("Times New Roman", 12);
+            using TestListBox listBox = new TestListBox();
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs:1390]:
```suggestion
            using Font font = new("Times New Roman", 12);
            using TestListBox listBox = new();
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs [hunk 8]
```diff
@@ -1382,6 +1382,30 @@ public static IEnumerable<object[]> HorizontalExtent_Set_TestData()
             }
         }
 
+        [WinFormsFact]
+        public void ListBox_Height_IsEqualDefaultHeight_IfDefaultFontIsChanged()
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs:1386]:
```suggestion
        public void ListBox_ItemHeight_EqualsDefaultFontHeight_IfDefaultFontIsChanged()
```

**Review by dmitrii-drobotov — COMMENTED** [review]
A few more nits, otherwise looks good!

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs (hunk 6 above)
Comment (dkazennov) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs:1398]: Yes. Now than you've pointed it out, I understood that I expected this class to be more useful.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 9]
```diff
@@ -586,7 +593,6 @@ public bool IntegralHeight
         ///  the height of an item in an owner-draw list box.
         /// </summary>
         [SRCategory(nameof(SR.CatBehavior))]
-        [DefaultValue(DefaultItemHeight)]
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:589]: I see this as a not observable change because we replace functionality with SholdSerialize and reset methods. https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/breaking-change-rules.md#attributes
@dreddy-work @JeremyKuhne  - thoughts?

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 10]
```diff
@@ -389,6 +395,8 @@ public bool UseCustomTabOffsets
             }
         }
 
+        private static int DefaultListBoxItemHeight => _defaultListBoxItemHeight;
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:398]: I think the property can be removed now, and every usage updated to just `_defaultListBoxItemHeight`, I don't see a reason to have both.

File: docs/list-of-diagnostics.md [hunk 11]
```diff
@@ -38,6 +38,8 @@ The acceptance criteria for adding an obsoletion includes:
 |  __`WFDEV001`__ | Casting to/from IntPtr is unsafe, use `ResultInternal`. |
 |  __`WFDEV002`__ | `DomainUpDown.DomainUpDownAccessibleObject` is no longer used to provide accessible support for `DomainUpDown` controls. Use `ControlAccessibleObject` instead. |
 |  __`WFDEV003`__ | `DomainUpDown.DomainItemAccessibleObject` is no longer used to provide accessible support for `DomainUpDown` items. |
+|  __`WFDEV004`__ | `ListBox.DefaultItemHeight` constant is no longer used as the default item height. 
+Default item height is now scaled according to the application default font. |
```
[the remark this answers is not in the document]
Comment (dkazennov) [reply docs/list-of-diagnostics.md:42]: Yes, thanks a lot! Missed this issue.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 12]
```diff
@@ -62,8 +67,9 @@ public partial class ListBox : ListControl
         private SelectedIndexCollection? _selectedIndices;
         private ObjectCollection? _itemsCollection;
 
-        private int _itemHeight = DefaultItemHeight;
+        private int _itemHeight = DefaultListBoxItemHeight;
         private int _columnWidth;
+        private static int _defaultListBoxItemHeight = DefaultFont.Height;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:72]: Are we guaranteed that `DefaultFont` is already initialized at this point?
It would be safer to do
```
private static int _defaultListBoxItemHeight = -1;

private static int DefaultListBoxItemHeight
{
    get
    {
        if (_defaultListBoxItemHeight == -1)
        {
            _defaultListBoxItemHeight = DefaultFont.Height;
        }
        return _defaultListBoxItemHeight;    
    }
}

Comment (dkazennov) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:72]: Yes, that's a good point!

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs (hunk 10 above)
Comment (dkazennov) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs]: So I've decided to keep the property (see below).

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 13]
```diff
@@ -62,8 +67,9 @@ public partial class ListBox : ListControl
         private SelectedIndexCollection? _selectedIndices;
         private ObjectCollection? _itemsCollection;
 
-        private int _itemHeight = DefaultItemHeight;
+        private int _itemHeight = DefaultListBoxItemHeight;
         private int _columnWidth;
+        private static int _defaultListBoxItemHeight = -1;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:72]: Please add a `s_` prefix per .editorconfig rules to this static field name.

**Comment by dkazennov** [comment]
Implemented final nit and rebased this fix on the most recent main as well.

**Review by dmitrii-drobotov — APPROVED** [review]

**Review by Tanya-Solyanik — APPROVED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs (hunk 9 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:589]: I am not completely sure how attribute default value is accessed without accessing `Get` property. But given the backing field is replaced with new default, looks safe to me.

**Review by dreddy-work — APPROVED** [review]

**Comment by dreddy-work** [comment]
@dkazennov / @Tanya-Solyanik , can you please create a doc bug here: https://github.com/dotnet/docs/issues, before merging this?

**Comment by Tanya-Solyanik** [comment]
@dreddy-work  - this should not be merged, it's waiting on the API review

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs (hunk 9 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:589]: Here is the get property - https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/ComponentModel/DefaultValueAttribute.cs#L232

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:589]: The change will be visible in application that are enumerating attributes on properties of this class. The purpose of this attribute is to be consumed by the TypeDescriptor, but since this is a public surface, anybody can read it.

**Comment by Tanya-Solyanik** [comment]
Breaking change doc for the obsoletion of the DefaultItemHeight contant https://github.com/dotnet/docs/issues/33953

**Comment by Tanya-Solyanik** [comment]
breaking change doc that covers all attributes removals that follow this pattern - https://github.com/dotnet/docs/issues/33954

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 14]
```diff
@@ -52,6 +52,11 @@ public partial class ListBox : ListControl
         ///  The default item height for an owner-draw ListBox. The ListBox's non-ownerdraw
         ///  item height is 13 for the default font on Windows.
         /// </summary>
+        [Obsolete(
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:55]: @dkazennov  - could you please move this obsoletion into a dedicated PR, We need to provide a replacement API. THe rest seems to be ready to go.

**Comment by dkazennov** [comment]
Obsoletion of the ListBox.DefaultItemHeight was moved to #8640.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs (hunk 14 above)
Comment (dkazennov) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs]: > @dkazennov - could you please move this obsoletion into a dedicated PR, We need to provide a replacement API. THe rest seems to be ready to go.

The new dedicated PR is #8640

**Review by Tanya-Solyanik — APPROVED** [review]
